TASK: Disallow disabling tethered nodes

nezaniel opened the pull request:
Resolves: #4821

With https://github.com/neos/neos-development-collection/pull/5496 node disabling was turned into a complete neos concern.
We discussed the possibility to constraint disabling when it was still a core concern. Now with the core being completely unaware of the disabled tag, we introduce a neos side high level constraint by leveraging a command hook `onBeforeHandle`:

```php
public function onBeforeHandle(CommandInterface $command): CommandInterface
{
   if (!$command instanceof TagSubtree || !$command->tag->equals(NeosSubtreeTag::disabled())) {
       return $command;
   }

   $nodeAggregate = $this->contentGraphReadModel->getContentGraph($command->workspaceName)->findNodeAggregateById($command->nodeAggregateId);
   if ($nodeAggregate?->classification->isTethered()) {
       throw new NodeAggregateIsTethered(
           sprintf('Node aggregate "%s" is classified as tethered and must not be disabled.', $nodeAggregate->nodeAggregateId->value),
           1741161426
       );
   }

   return $command;
}
```

As we'd like to support the general feature to tag tethered nodes with custom tags from the cr perspective, no constraints are enforced by the core and a test has been added to show its possible to tag tethered nodes arbitrary.

This change sets the foundation for https://github.com/neos/neos-development-collection/pull/5479 which additionally will prevent soft removing tethered nodes - a restriction which is definitely required in neos as hard removal of tethered nodes is impossible and so must soft removal too.

The argumentation that a subtree must be valid to the current schema at any point - basically why we don't want to be able to hide tethered nodes - has two technical flaws:
For one it can never really be relied upon that the subtree matches the node type schema as this can easily diverge and will only be in sync after structure adjustments.
Secondly if we still allow arbitrary tags on core level, one is still allowed to exclude these tags via `VisibilityConstraint` and runs into the same problem as if a node was disabled if one doesnt have the permissions to see a node. Though it can be argued that this case is more expected than hiding the `main` content collection and having the rendering fail because of that unexpectedly.

Also on the argumentation that disabling tethered nodes is a feature, we outvoted bastians [concerns](https://github.com/neos/neos-ui/pull/3004#issuecomment-1156175008) for now.

Due to the minor flaws in the concept and opposition its not totally unlikely that this decision might be reverted and in combination with adjustments in the neos ui allowed to disable tethered nodes.
Though for the time being - the neos 9 release - we wont change this behaviour from Neos 8 and this pr will just enforce that also on the server side.

Further this change introduces a `SubtreeTagging` namespace in Neos' `Domain` and puts there the new `NeosSubtreeTag` and also the shown Neos subtreetag constraint check.

mhsdesign commented:
❤️

mhsdesign commented:
The branch is a bit broken now, i presume as it was dependant on the workspacePr it got rebased or sth?

mhsdesign commented:
i tried to cherry pick the changes `f3614a755512c22de3c9085afda31db9ef098167` (original commit in details)

<details><summary>Details</summary>
<p>


```diff
From f3614a755512c22de3c9085afda31db9ef098167 Mon Sep 17 00:00:00 2001
From: Bernhard Schmitt <schmitt@sitegeist.de>
Date: Wed, 13 Mar 2024 09:43:13 +0100
Subject: [PATCH] 4821 - Disallow disabling tethered nodes

---
 ...ableNodeAggregate_ConstraintChecks.feature | 22 ++++++++++++++-----
 .../Feature/NodeDisabling/NodeDisabling.php   |  1 +
 2 files changed, 17 insertions(+), 6 deletions(-)

diff --git a/Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/06-NodeDisabling/01-DisableNodeAggregate_ConstraintChecks.feature b/Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/06-NodeDisabling/01-DisableNodeAggregate_ConstraintChecks.feature
index d41691053f..91e3033a1b 100644
--- a/Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/06-NodeDisabling/01-DisableNodeAggregate_ConstraintChecks.feature
+++ b/Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/06-NodeDisabling/01-DisableNodeAggregate_ConstraintChecks.feature
@@ -11,7 +11,11 @@ Feature: Constraint checks on node aggregate disabling
       | language   | de, gsw, en | gsw->de, en     |
     And using the following node types:
     """yaml
-    'Neos.ContentRepository.Testing:Document': []
+    'Neos.ContentRepository.Testing:Tethered': []
+    'Neos.ContentRepository.Testing:Document':
+      childNodes:
+        tethered:
+          type: 'Neos.ContentRepository.Testing:Tethered'
     """
     And using identifier "default", I define a content repository
     And I am in content repository "default"
@@ -30,8 +34,8 @@ Feature: Constraint checks on node aggregate disabling
       | nodeTypeName            | "Neos.ContentRepository:Root" |
     And the graph projection is fully up to date
     And the following CreateNodeAggregateWithNode commands are executed:
-      | nodeAggregateId | nodeTypeName                            | parentNodeAggregateId | nodeName |
-      | sir-david-nodenborough  | Neos.ContentRepository.Testing:Document | lady-eleonode-rootford        | document |
+      | nodeAggregateId | nodeTypeName                            | parentNodeAggregateId | nodeName | tetheredDescendantNodeAggregateIds |
+      | sir-david-nodenborough  | Neos.ContentRepository.Testing:Document | lady-eleonode-rootford        | document | {"tethered": "nodewyn-tetherton"}                   |
 
   Scenario: Try to disable a node aggregate in a non-existing content stream
     When the command DisableNodeAggregate is executed with payload and exceptions are caught:
@@ -58,6 +62,13 @@ Feature: Constraint checks on node aggregate disabling
       | nodeVariantSelectionStrategy | "allVariants"                          |
     Then the last command should have thrown an exception of type "NodeAggregateCurrentlyDoesNotExist"
 
+  Scenario: Try to disable a tethered node aggregate
+    When the command DisableNodeAggregate is executed with payload and exceptions are caught:
+      | Key                          | Value                                  |
+      | nodeAggregateId              | "nodewyn-tetherton"                    |
+      | nodeVariantSelectionStrategy | "allVariants"                          |
+    Then the last command should have thrown an exception of type "NodeAggregateIsTethered"
+
   Scenario: Try to disable an already disabled node aggregate
     Given the command DisableNodeAggregate is executed with payload:
       | Key                          | Value                                  |
@@ -72,14 +83,13 @@ Feature: Constraint checks on node aggregate disabling
       | nodeAggregateId      | "sir-david-nodenborough"               |
       | coveredDimensionSpacePoint   | {"language": "de"}                                     |
       | nodeVariantSelectionStrategy | "allVariants"                          |
-    Then I expect exactly 4 events to be published on stream with prefix "ContentStream:cs-identifier"
-    And event at index 3 is of type "NodeAggregateWasDisabled" with payload:
+    Then I expect exactly 5 events to be published on stream with prefix "ContentStream:cs-identifier"
+    And event at index 4 is of type "NodeAggregateWasDisabled" with payload:
       | Key                          | Expected                               |
       | contentStreamId              | "cs-identifier"                        |
       | nodeAggregateId              | "sir-david-nodenborough"               |
       | affectedDimensionSpacePoints | [{"language":"de"},{"language":"gsw"}] |
 
-
   Scenario: Try to disable a node aggregate in a non-existing dimension space point
     When the command DisableNodeAggregate is executed with payload and exceptions are caught:
       | Key                          | Value                                  |
diff --git a/Neos.ContentRepository.Core/Classes/Feature/NodeDisabling/NodeDisabling.php b/Neos.ContentRepository.Core/Classes/Feature/NodeDisabling/NodeDisabling.php
index 84bcfccd4e..9ae2372b96 100644
--- a/Neos.ContentRepository.Core/Classes/Feature/NodeDisabling/NodeDisabling.php
+++ b/Neos.ContentRepository.Core/Classes/Feature/NodeDisabling/NodeDisabling.php
@@ -60,6 +60,7 @@ private function handleDisableNodeAggregate(
             $nodeAggregate,
             $command->coveredDimensionSpacePoint
         );
+        $this->requireNodeAggregateToBeUntethered($nodeAggregate);
 
         if ($nodeAggregate->disablesDimensionSpacePoint($command->coveredDimensionSpacePoint)) {
             // already disabled, so we can return a no-operation.
```

</p>
</details> 


but it seems to need further adjustments and conception after the subtree tags

mhsdesign commented:
How do we want to continue here?

nezaniel commented:
I'd say we allow general tagging but not disabling on tethered nodes for now

mhsdesign commented:
Today we discussed we disallow `TagSubtree(disable)` `DisableNodeAggregate` for tethered nodes.

mhsdesign committed:
TASK: Move subtree tagging constraint checks into own feature

these checks where previously last changed via https://github.com/neos/neos-development-collection/pull/5357

mhsdesign reviewed (approved):
Makes now especially sense as we will also prevent tagging a tethered node as "removed" via https://github.com/neos/neos-development-collection/pull/5479 -> where the constraints will be on the Neos side though ...

In case we dont want to go through with this change at core level, we could still decide to move it into neos as concept, similarly like we did with removed.

mhsdesign commented:
On the other hand, @kitsunet and me wondered why the node disabling is still a core concern at all it ought to be deprecated: https://github.com/neos/neos-development-collection/pull/5496

In _that_ case, we must really go the way to impose high level constraints via neos checks and should not do that in the core.

mhsdesign committed:
TASK: Introduce basic core test that tagging tethered and root nodes works

mhsdesign committed:
TASK: Add Neos subtreeTagging constraint checks

that disallows disabling tethered nodes

mhsdesign committed:
TASK: Colocate new NeosSubtreeTag and NeosVisibilityConstraints in `SubtreeTagging` namespace

mhsdesign commented on Neos.Neos/Classes/Domain/SubtreeTagging/NeosSubtreeTag.php:
i moved the new factories into this new `SubtreeTagging` namespace as both arent really a `Service` and its good to have things colocated.

The soft removal garbage collection can then be placed into `SubtreeTagging\SoftRemoval`

Hope you agree with that change as well @kitsunet ? ^^ Even though it causes a little diff for the just merged files ^^

kitsunet replied:
Nice
